### ISSUE TITLE
Accuracy degradation on bosch dataset

### ISSUE THREAD

**Issue by razdoburdin** [dmlc/xgboost#11198]
Hi,
I have observed the accuracy degradation for `bosch-production-line-performance` dataset after this PR https://github.com/dmlc/xgboost/pull/10298.
So, for commit ab8aa6f126156ad7fffb56a4a9baae06c0228f85 the accuracy is **0.844** for 100 estimators, but for ab228cc78d054fc6652728fdcfc229ee0e4de822 it drops to **0.762**.

**Comment by trivialfis** [comment]
Could you please share your code for the benchmark? Is the accuracy from validation dataset or from the training dataset?

**Comment by trivialfis** [comment]
Initial result from my test:
```
* master
[99]    Train-logloss:0.02704   Train-auc:0.89945       Valid-logloss:0.03350   Valid-auc:0.71049
Train acc 0.9946429979387481
Valid acc 0.9942825668976844
Duration: 38.64595603942871

* 2.1
[99]    Train-logloss:0.02409   Train-auc:0.91872       Valid-logloss:0.03095   Valid-auc:0.71511
Train acc 0.9948367330847816
Valid acc 0.9943873189226086
Duration: 40.097050189971924
```

**Comment by razdoburdin** [comment]
Sorry for a log delay:
Here is a small reproducer:

```
import numpy as np
import xgboost as xgb

X_train = np.load('./data/bosch_x_train.npy')
X_test = np.load('./data/bosch_x_test.npy')
y_train = np.load('./data/bosch_y_train.npy')
y_test = np.load('./data/bosch_y_test.npy')

dtrain = xgb.DMatrix(X_train, y_train)
dtest = xgb.DMatrix(X_test, y_test)

params = {
    'booster': 'gbtree',
    'device': 'cpu',
    'tree_method': 'hist',
    'objective': 'binary:logistic',
    'max_depth': 8,
    'learning_rate': 0.1,
    'scale_pos_weight': 173.63348001466812
}

booster = xgb.train(params, dtrain, 100)

y_pred = booster.predict(dtrain)
y_pred = (y_pred >= 0.5).astype(np.int32)
print(f'train acc = {1 - np.sum(np.abs(y_pred - y_train)) / y_train.size}')
```

I have double checked the accuracy for current `master` branch and for `release_2.1.0`, and observed the degradation from **0.844** to **0.762**.

**Comment by trivialfis** [comment]
Hmm, so there's a scale pos weight. The mean-based initialization method doesn't consider that.

**Comment by trivialfis** [comment]
Using the numeric portion of the data:
```
r2.1
train acc = 0.8310880618916036
master
train acc = 0.8473947558050834
```
Will address it.
